dashboard/app/controllers/api_controller.rb

```diff
@@ -546,6 +547,120 @@ def assessments_structure
     render json: assessments
   end
 
+  # Return a hash by student_id for all students in a section. The value is a hash by
+  # script_level_id, with information on the student responses for that assessment.
+  # Currently, very similar logic to section_assessments, but will change as we refactor. The format
+  # of the results is different enough that I'm duplicating some code it for now.
+  # TODO(caleybrock): currently only used in internal experiment, must add controller tests.
+  def assessments_responses
+    section = load_section
+    script = load_script(section)
+
+    responses_by_student = {}
+
+    level_group_script_levels = script.script_levels.includes(:levels).where('levels.type' => 'LevelGroup')
+
+    section.students.each do |student|
+      # Initialize student hash
+      student_hash = {
+        student: {id: student.id, name: student.name}
+      }
+      responses_by_level_group = {}
+
+      level_group_script_levels.each do |script_level|
+        next unless script_level.long_assessment?
+
+        # Don't allow somebody to peek inside an anonymous survey using this API.
+        next if script_level.anonymous?
+
+        # Get the UserLevel for the last attempt.  This approach does not check
+        # for the script and so it'll find the student's attempt at this level for
+        # any script in which they have encountered that level.
+        last_attempt = student.last_attempt_for_any(script_level.levels)
+
+        # Get the LevelGroup itself.
+        level_group = last_attempt.try(:level) || script_level.oldest_active_level
+
+        # Get the response which will be stringified JSON.
+        response = last_attempt.try(:level_source).try(:data)
+
+        next unless response
+
+        # Parse the response string into an object.
+        response_parsed = JSON.parse(response)
+
+        # Summarize some key data.
+        multi_count = 0
+        multi_count_correct = 0
+
+        # And construct a listing of all the individual levels and their results.
+        level_results = []
+
+        level_group.levels.each do |level|
+          if level.is_a? Multi
+            multi_count += 1
+          end
+
+          level_response = response_parsed[level.id.to_s]
+
+          level_result = {}
+
+          if level_response
+            case level
```

Comment:
I thought that some of this makes more sense on a model, but I wasn't sure which one since it has user data and level data. Open to suggestions since this was copied and pasted from the current API.

Reply:
I actually don't know what the Rails-y answer is for this (@madelynkasula? @aoby?).  You're kind of just generating a view on some data based on multiple models, and my instinct would be to delegate that work somehow, and reduce the controller's job to authorization and reading request parameters.

```rb
def assessments_responses
  section = load_section
  script = load_script(section)
  render json: AssessmentsResponsesBuilder.new(section, script).responses_by_student
end
```

Again, not sure that's idiomatic Rails.

Reply:
yep, i believe that's idiomatic rails--["skinny controller, fat model"](http://weblog.jamisbuck.org/2006/10/18/skinny-controller-fat-model) is usually used to refer to this. even if there are two models in the mix, and moving the logic would mean models having to talk to each other, that can be okay as long as the related logic is isolated as much as possible.

an example of this would be [here](https://github.com/code-dot-org/code-dot-org/blob/12161960163c9829d6e288555648e1198977d23f/dashboard/app/models/course.rb#L242-L244). a `course` needs to know if any `experiment`s are enabled for a `user` (3 models), but that logic is isolated to a single method rather than getting tangled up in other model logic

Reply:
Agreed. This seems like an awful lot of logic for a controller. Move the complex logic into a model if you can. If it's not appropriate to a single model (which could mean from that model's perspective even if it includes associations, as @madelynkasula suggested), then it can go in a common `lib` module for standalone code or in a model concern which can be shared between models. When the logic is in the model (or lib) it can also be tested in isolation.

Reply:
The only difference I'd call out from Brad's suggestion is to use an ActiveModelSerializer instead of a custom builder class to format the response JSON.